src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj

```diff
@@ -483,22 +483,39 @@ This package is an internal implementation of the .NET Core SDK and is not meant
 
     <PropertyGroup>
       <CompositeTargetDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', 'Composite', '$(Configuration)'))</CompositeTargetDir>
+      <CompositeFileName>full-composite.r2r</CompositeFileName>
+      <CompositeResponseFile>$(CompositeTargetDir)\$(CompositeFileName).rsp</CompositeResponseFile>
+      <CompositeArgs>@$(CompositeResponseFile)</CompositeArgs>
       <ManagedAssetsFullPath>$(RuntimePackageRoot)$(ManagedAssetsPackagePath)</ManagedAssetsFullPath>
       <NativeAssetsFullPath>$(RuntimePackageRoot)$(NativeAssetsPackagePath)</NativeAssetsFullPath>
-
-      <BaseAssemblies>$(ManagedAssetsFullPath)\*.dll</BaseAssemblies>
-
-      <CompositeArgs>--composite</CompositeArgs>
-      <CompositeArgs>$(CompositeArgs) --targetarch:$(TargetArchitecture)</CompositeArgs>
-      <CompositeArgs>$(CompositeArgs) --targetos:$(Crossgen2TargetOs)</CompositeArgs>
-      <CompositeArgs>$(CompositeArgs) $(BaseAssemblies)</CompositeArgs>
-      <CompositeArgs>$(CompositeArgs) $(TargetDir)*.dll</CompositeArgs>
-      <CompositeArgs>$(CompositeArgs) --out:&quot;$(CompositeTargetDir)full-composite.r2r.dll&quot;</CompositeArgs>
+      <PartialCompositeAssemblyListPath>$(MSBuildThisFileDirectory)\PartialCompositeAssemblyList.txt</PartialCompositeAssemblyListPath>
     </PropertyGroup>
 
     <RemoveDir Directories="$(CompositeTargetDir)" />
     <MakeDir Directories="$(CompositeTargetDir)" />
 
+    <!-- Read list of assemblies that should generate native code in the full composite image. -->
+    <ReadLinesFromFile File="$(PartialCompositeAssemblyListPath)">
+      <Output TaskParameter="Lines" ItemName="PartialCompositeAssemblyNames" />
+    </ReadLinesFromFile>
+
+    <ItemGroup>
+      <BaseAssemblies Include="$(ManagedAssetsFullPath)\*.dll" />
+      <BaseAssemblies Include="$(TargetDir)\*.dll" />
+      <PartialCompositeAssemblyPaths Include="$(ManagedAssetsFullPath)\%(PartialCompositeAssemblyNames.Identity).dll" />
+      <PartialCompositeAssemblyPaths Include="$(TargetDir)\%(PartialCompositeAssemblyNames.Identity).dll" />
+      <PartialCompositeAssemblyPaths Remove="%(PartialCompositeAssemblyPaths.Identity)" Condition="!Exists('%(Identity)')" />
+      <UnrootedAssemblyPaths Include="@(BaseAssemblies)" />
```

Comment:
I'm confused. What is this `UnrootedAssemblyPaths`? It seems to be a list of the assemblies that will not be part of the composite. Why do we need it?

Reply:
The logic I'm using here is slightly different than the measurements you were doing last year. In your implementation you were creating the composite image from just a subset of the framework / ASP.NET assemblies and the remaining assemblies were left in their original shipping (R2R-compiled) form. This approach uses the "full filtered composite" where we make a composite comprising all framework and ASP.NET assemblies but not all of them generate R2R code into the image; in particular, for "unrooted inputs", we don't mark all their methods as roots - they get compiled if they're reachable from another assembly via dependency analysis, they're just not compiled automatically. This filtering is what lets us reduce the total publishing size while keeping startup time on par with default R2R publishing - for some assemblies that are rarely used we drop the R2R code altogether in this scenario.

Reply:
Ok thanks for the explanation Tomas. Then I think this is ready to be merged.